File: src/main/java/cz/vhromada/catalog/web/validator/constraints/Years.java
package cz.vhromada.catalog.web.validator.constraints;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import javax.validation.Constraint;
import javax.validation.Payload;

import cz.vhromada.catalog.web.validator.YearsValidator;

/**
 * An annotation represents years constraint.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = YearsValidator.class)
@Documented
public @interface Years {

    String message() default "{cz.vhromada.catalog.validator.Years.message}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/cz/vhromada/catalog/web/validator/constraints/Years.kt
package cz.vhromada.catalog.web.validator.constraints

import cz.vhromada.catalog.web.validator.YearsValidator
import javax.validation.Constraint
import javax.validation.Payload
import kotlin.reflect.KClass

/**
 * An annotation represents years constraint.
 *
 * @author Vladimir Hromada
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [YearsValidator::class])
@MustBeDocumented
annotation class Years(

        /**
         * Returns error message template.
         *
         * @return error message template
         */
        val message: String = "{cz.vhromada.catalog.validator.Years.message}",

        /**
         * Returns groups constraint belongs to.
         *
         * @return groups constraint belongs to
         */
        val groups: Array<KClass<*>> = [],

        /**
         * Returns payload associated to constraint.
         *
         * @return payload associated to constraint
         */
        @Suppress("unused")
        val payload: Array<KClass<out Payload>> = [])
